slkspec/core.py

```diff
@@ -1,129 +1,354 @@
+    **kwargs:
+        Additional keyword arguments passed to the open file descriptor method.
+
+    Example
+    -------
+
+    Use fsspec to open data stored on tape, temporary data will be downloaded
+    to a central scratch folder:
+
+    ::
+
+        import ffspec
+        import xarray as xr
+
+        url = fsspec.open("slk:////arch/bb1203/data.nc",
+                          slk_cache="/scratch/b/b12346").open()
+        dset = xr.open_dataset(url)
+
+    """
+
+    write_msg: str = "Write mode is not suppored"
+    """Error message that is thrown if the files are attempted to be opened
+    in any kind of write mode."""
+
+    def __init__(
+        self,
+        url: str,
+        local_file: str,
+        *,
+        override: bool = True,
+        mode: str = "rb",
+        touch: bool = True,
+        file_permissions: int = 0o3777,
+        _lock: threading.Lock = _retrieval_lock,
+        _file_queue: Queue[Tuple[str, str]] = FileQueue,
+        **kwargs: Any,
+    ):
+        if not set(mode) & set("r"):  # The mode must have a r
+            raise NotImplementedError(self.write_msg)
+        if "b" not in mode:
+            kwargs.setdefault("encoding", "utf-8")
+        self._file = str(Path(local_file).expanduser().absolute())
+        self._url = str(url)
+        self.touch = touch
+        self.file_permissions = file_permissions
+        self._order_num = 0
+        self._file_obj: Optional[IO[Any]] = None
+        self._lock = _lock
+        self.kwargs = kwargs
+        self.mode = mode
+        self.newlines = None
+        self.error = "strict"
+        self.encoding = kwargs.get("encoding")
+        self.write_through = False
+        self._file_queue = _file_queue
+        print(self._file)
+        with _lock:
+            if not Path(self._file).exists() or override:
+                self._file_queue.put((self._url, str(Path(self._file).parent)))
+            elif Path(self._file).exists():
+                if self.touch:
+                    Path(self._file).touch()
+                self._file_obj = open(self._file, mode, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Get the file for the SLKFile object."""
+        if self._file_obj is not None:
+            return self._file
+        return self._url
+
+    def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
+        """Get items from the tape archive."""
+
+        retrieval_requests: Dict[Path, List[str]] = defaultdict(list)
+        logger.debug("Retrieving %i items from tape", len(retrieve_files))
+        for inp_file, out_dir in retrieve_files:
+            retrieval_requests[Path(out_dir)].append(inp_file)
+        for output_dir, inp_files in retrieval_requests.items():
+            output_dir.mkdir(parents=True, exist_ok=True, mode=self.file_permissions)
+            logger.debug("Creating slk query for %i files", len(inp_files))
+            search_str = pyslk.slk_search(pyslk.slk_gen_file_query(inp_files))
```

Comment:
This is not yet taking advantage of combining retrievals, right?
I imagine something like:
```python
pyslk.slk_search(pyslk.slk_gen_file_query(retrieval_requests.values()))
```
so that all files that are currently in the queue are gathered in a single search.

Coming back to my catalog example:

```python
from intake import open_catalog
cat=open_catalog("https://raw.githubusercontent.com/observingClouds/tape_archive_index/main/catalog.yml")
ds=cat["EUREC4A_ICON-LES_highCCN_DOM02_surface_native"].to_dask()
ds[["u_10m","v_10m"]].isel(time=slice(0,1)).mean().compute() # this requests two tar files from tape
```
The log file currently shows:
```
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
/scratch/m/m300408/retrieval/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface/EUREC4A_ICON-LES_highCCN_DOM02_surface_native.v_10m.007.tar
slk search '{"$and":[{"path":{"$gte":"/arch/mh0010/m300408/experiments/EUREC4A/highCCN/output/DOM02/surface","$max_depth":1}},{"resources.name":{"$regex":"EUREC4A_ICON-LES_highCCN_DOM02_surface_native.u_10m.006.tar"}}]}'
```
So all files are requested, but only the first one "searched" and retrieved and eventually the next file will be requested separately.

I can merge this though and we can work on the remaining issues/features in finer increments/PRs. This PR works now similar to the initial version, but has of course also advanced already by a lot 👍

Reply:
From a code perspective, the queue with the thread lock should be able to do this but @neumannd suggested this retrieval strategy. Because currently, we can only retrieve files residing in the same directory at once.

Suppose that you have:

```
/arch/foo/file1.txt
/arch/foo/file2.txt
/arch/foo/file3.txt
/arch/bar/file1.txt
/arch/bar/file2.txt
/arch/bar/file3.txt
```

Then everything that is in `/arch/foo` would be retrieved once while `/arch/bar` would be processed in a separate request. That's why the two loops. The first gathers the unique dirs `/arch/foo` `/arch/bar` etc and then requests the files in those dirs.

What we maybe could do is process retrievals in a multi-process/thread pool. @neumannd what do you think about doing the `pysl.slk_search` and `pyslk.slk_retrieve` in parallel (essentially the second loop).

Reply:
Note that in my example above the files are all residing in the same directory.

Reply:
Ok I'll take a look at this, to see whats wrong.

Reply:
@antarcticrainforest With slk 3.3.67 the `slk retrieve SEARCH_ID /target` will recreate the whole path of each found file in `/target`. I forgot to mention that. Thus, if search_id 123456 found the files from the example above and if we were doing `slk retrieve 123456 /target` then we would get folders `/target/arch/foo` and `/target/arch/bar` (and the files would be copied into them accordingly).

Reply:
@observingClouds I spend half an hour poking around and I think the problem doesn't come from the queue implementation but is rather on `preffs` side. Before we add additional layers of complexity, would you be able to provide a catalogue that points to traditional netcdf files (with `slk` as prefix ) residing in the same directory? I would then run some tests. To confirm whether or not I will have to change the architecture. Thanks.

Reply:
@antarcticrainforest, you might be right that the error is further upstream. Possibly `fsspec` itself. I need to look further into this. There might be an issue with url-chaining.
A working example with netCDF would be:
```python
import xarray as xr
import fsspec
m=fsspec.open_files(['slk:///arch/mh0010/m300408/showcase/dataset*.nc'])
with m3 as f:
    xr.open_mfdataset(f, engine="h5netcdf")
```

```python
m=fsspec.open_files(['simplecache::slk:///arch/mh0010/m300408/showcase/dataset*.nc'])
```
with an additional protocol, e.g. `simplecache`, in the url, makes the retrievals "serial" again.

Reply:
@antarcticrainforest despite the example above, I think slkspec needs to be derived from the `AsyncFileSystem` class to support joined retrieval in everyday situations. Here is another example that does not have url chaining, but still fails:
```python
import xarray
xr.open_mfdataset("slk:///arch/mh0010/m300408/showcase/dataset.zarr", engine="zarr")
```

Reply:
I cannot completely test my example, because I get the error `PySlkException: pyslk.slk_search: search command is disabled`

Reply:
@observingClouds It's fixed now. Sorry for that. Sometimes wrong config files are deployed the levante nodes.

Reply:
moved to #10